lib/spack/spack/container/__init__.py

```diff
@@ -0,0 +1,73 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+"""Package that provides functions and classes to
+generate container recipes from a Spack environment
+"""
+import warnings
+
+import spack.schema.env as env
+import spack.util.spack_yaml as syaml
+from .writers import recipe
+
+__all__ = ['validate', 'recipe']
+
+
+def validate(configuration_file):
+    """Validate a Spack environment YAML file that is being used to generate a
+    recipe for a container.
+
+    Since a few attributes of the configuration must have specific values for
+    the container recipe, this function returns a sanitized copy of the
+    configuration in the input file. If any modification is needed, a warning
+    will be issued.
+
+    Args:
+        configuration_file (str): path to the Spack environment YAML file
+
+    Returns:
+        A sanitized copy of the configuration stored in the input file
+    """
+    import jsonschema
+    with open(configuration_file) as f:
+        config = syaml.load(f)
+
+    # The "container" subsection is mandatory
+    msg = '"container" subsection missing in "{0}"'
+    assert 'container' in config['spack'], msg.format(configuration_file)
+
+    # Remove attributes that are not needed / allowed in the
+    # container recipe
+    for subsection in ('cdash', 'gitlab_ci', 'modules'):
+        if subsection in config['spack']:
+            msg = ('the subsection "{0}" in "{1}" is not used when generating'
+                   ' container recipes and will be discarded')
+            warnings.warn(msg.format(subsection, configuration_file))
+            config['spack'].pop(subsection)
+
+    # Set the default value of the concretization strategy to "together" and
+    # warn if the user explicitly set another value
+    config['spack'].setdefault('concretization', 'together')
+    if config['spack']['concretization'] != 'together':
+        msg = ('the "concretization" attribute of the environment is set '
+               'to "{0}" [the advised value is instead "together"]')
+        warnings.warn(msg.format(config['spack']['concretization']))
+
+    # Check if the install tree was explicitly set to a custom value and warn
+    # that it will be overridden
+    environment_config = config['spack'].get('config', {})
+    if environment_config.get('install_tree', None):
```

Comment:
Should users be able to configure where software is put in the container?

Reply:
I would say not at the moment, since there's probably no clear benefit (or at least I can't see it) on let users decide where to install things in a container. Leaving it fixed to a path we decide, for the time being, will simplify the choices users have to make when configuring the Spack environment.

Also, internally templates and corresponding code are already parametrized on these paths:

https://github.com/spack/spack/blob/b2d1a6da51a347110190f9ddc5ec9e7317331001/lib/spack/spack/container/writers/__init__.py#L86-L96

so if a use case comes up where it's crucial to have this extra flexibility we could easily add that in the future.